javascript-algorithms-and-data-structures/Learn-Intermediate-OOP-by-Building-a-Platformer-Game/script.js

```diff
@@ -0,0 +1,231 @@
+      if (this.position.y < 0) {
+        this.position.y = 0;
+        this.velocity.y = gravity;
+      }
+      this.velocity.y += gravity;
+    } else {
+      this.velocity.y = 0;
+    }
+    if (this.position.x < this.width) {
+      this.position.x = this.width;
+    }
+  }
+}
+
+class Platform {
+  constructor(x, y) {
+    this.position = {
+      x,
+      y,
+    };
+    this.width = 200;
+    this.height = 40;
+  }
+  draw() {
+    ctx.fillStyle = "#acd157";
+    ctx.fillRect(this.position.x, this.position.y, this.width, this.height);
+  }
+}
+
+class CheckPoint {
+  constructor(x, y) {
+    this.position = {
+      x,
+      y,
+    };
+    this.width = 40;
+    this.height = 70;
+  }
+  draw() {
+    ctx.fillStyle = "#f1be32";
+    ctx.fillRect(this.position.x, this.position.y, this.width, this.height);
+  }
+}
+
+const player = new Player();
+const platformPositions = [
+  { x: 500, y: 450 },
+  { x: 700, y: 400 },
+  { x: 850, y: 350 },
+  { x: 900, y: 350 },
+  { x: 1050, y: 150 },
+  { x: 2500, y: 450 },
+  { x: 2900, y: 400 },
+  { x: 3150, y: 350 },
+  { x: 3900, y: 450 },
+  { x: 4200, y: 400 },
+  { x: 4400, y: 200 },
+  { x: 4700, y: 150 },
+];
+
+const platforms = platformPositions.map(
+  (platform) => new Platform(platform.x, platform.y)
+);
+
+const checkpointPositions = [
+  { x: 1170, y: 80 },
+  { x: 2900, y: 330 },
+  { x: 4800, y: 80 },
+];
+const checkpoints = checkpointPositions.map(
+  (checkpoint) => new CheckPoint(checkpoint.x, checkpoint.y)
+);
+
+const animate = () => {
+  requestAnimationFrame(animate);
+
+  ctx.clearRect(0, 0, canvas.width, canvas.height);
+
+  platforms.forEach((platform) => {
+    platform.draw();
+  });
+
+  checkpoints.forEach((checkpoint) => {
+    checkpoint.draw();
+  });
+
+  player.update();
+
+  if (keys.rightKey.pressed && player.position.x < 400) {
+    player.velocity.x = 5;
+  } else if (keys.leftKey.pressed && player.position.x > 100) {
+    player.velocity.x = -5;
+  } else {
+    player.velocity.x = 0;
+
+    if (keys.rightKey.pressed) {
+      platforms.forEach((platform) => {
+        platform.position.x -= 5;
+      });
+      checkpoints.forEach((checkpoint) => {
+        checkpoint.position.x -= 5;
+      });
+    } else if (keys.leftKey.pressed) {
+      platforms.forEach((platform) => {
+        platform.position.x += 5;
+      });
+      checkpoints.forEach((checkpoint) => {
+        checkpoint.position.x += 5;
+      });
+    }
+  }
+
+  platforms.forEach((platform) => {
+    const collisionDetectionRules = [
+      player.position.y + player.height <= platform.position.y,
+      player.position.y + player.height + player.velocity.y >=
+        platform.position.y,
+      player.position.x >= platform.position.x - player.width / 2,
+      player.position.x <=
+        platform.position.x + platform.width - player.width / 3,
+    ];
+
+    if (collisionDetectionRules.every((rule) => rule)) {
+      player.velocity.y = 0;
+    }
+  });
```

Comment:
```suggestion
  platforms.forEach((platform) => {
    const collisionDetectionRules = [
      player.position.y + player.height <= platform.position.y,
      player.position.y + player.height + player.velocity.y >=
      platform.position.y,
      player.position.x >= platform.position.x - player.width / 2,
      player.position.x <=
      platform.position.x + platform.width - player.width / 3,
    ];
    console.log(player.position.y, player.height, platform.position.y)
    if (collisionDetectionRules.every((rule) => rule)) {
      player.velocity.y = 0;
      return;
    }
    if (
      player.position.x >= platform.position.x - player.width / 2 &&
      player.position.x <=
      platform.position.x + platform.width - player.width / 3 &&
      player.position.y + player.height >= platform.position.y &&
      player.position.y <= platform.position.y + platform.height
    ) {
      player.position.y = platform.position.y + player.height;
      player.velocity.y = gravity;
    }
  });
```

I feel like this suggestion could be cleaned up, but I'm struggling to see how. Essentially, the goal here is to detect bottom-border collisions and prevent players from being able to jump up through platforms.